<h1><center>Traiing the Iris Dataset with 3 ML Models</center></h1>

In [10]:
# Begin by importing all necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier # K-Nearest Neibour Model
from sklearn.naive_bayes import GaussianNB         # Gaussian Naive Bayes Model
from sklearn.svm import SVC                        # Support Vector Machine Model

In [11]:
data = pd.read_csv('iris.csv')
# Check the first 5 entries
print(data.head(5))

   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa


<h3>Define the features and labels in Pandas by slicing the data table and choosing certain rows/columns with iloc(). The slicing notation above selects every row and every column except the last column which contains the label of the species).<br />

In [12]:
data.drop('Id', axis=1, inplace=True)
print(data.head(5))
# Pandas ".iloc" expects row_indexer, column_indexer  
X = data.iloc[:,:-1].values
# Alternate way of selecting columns:
# X = data.iloc['SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm']
# Specify which column we want for the target/labels.  
y = data['Species']

   SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0            5.1           3.5            1.4           0.2  Iris-setosa
1            4.9           3.0            1.4           0.2  Iris-setosa
2            4.7           3.2            1.3           0.2  Iris-setosa
3            4.6           3.1            1.5           0.2  Iris-setosa
4            5.0           3.6            1.4           0.2  Iris-setosa


<h3>Next split the data into training and testing sets using the train_test_split() function</h3>

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=27)

In [14]:
print(X_train)  
print(y_train)

[[6.3 2.7 4.9 1.8]
 [5.  3.5 1.3 0.3]
 [6.4 2.7 5.3 1.9]
 [6.2 2.9 4.3 1.3]
 [6.7 3.1 4.7 1.5]
 [4.4 3.  1.3 0.2]
 [6.8 2.8 4.8 1.4]
 [6.3 2.5 5.  1.9]
 [5.8 2.7 3.9 1.2]
 [4.8 3.1 1.6 0.2]
 [4.6 3.4 1.4 0.3]
 [6.3 2.9 5.6 1.8]
 [5.9 3.2 4.8 1.8]
 [7.2 3.2 6.  1.8]
 [4.6 3.6 1.  0.2]
 [6.  2.9 4.5 1.5]
 [6.8 3.  5.5 2.1]
 [7.4 2.8 6.1 1.9]
 [6.1 2.9 4.7 1.4]
 [5.  3.5 1.6 0.6]
 [5.4 3.7 1.5 0.2]
 [5.1 3.5 1.4 0.2]
 [5.5 2.4 3.8 1.1]
 [5.2 3.5 1.5 0.2]
 [6.3 2.3 4.4 1.3]
 [5.7 3.8 1.7 0.3]
 [6.7 3.1 5.6 2.4]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.3]
 [6.5 2.8 4.6 1.5]
 [5.8 2.7 5.1 1.9]
 [6.2 2.8 4.8 1.8]
 [5.7 3.  4.2 1.2]
 [7.1 3.  5.9 2.1]
 [6.7 3.3 5.7 2.5]
 [5.5 2.5 4.  1.3]
 [6.4 3.1 5.5 1.8]
 [6.4 2.8 5.6 2.2]
 [4.3 3.  1.1 0.1]
 [6.9 3.1 5.4 2.1]
 [5.  3.2 1.2 0.2]
 [6.5 3.  5.2 2. ]
 [6.7 3.3 5.7 2.1]
 [5.1 3.8 1.9 0.4]
 [4.9 2.4 3.3 1. ]
 [5.1 3.8 1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.8 4.  1.2 0.2]
 [5.8 2.7 4.1 1. ]
 [5.4 3.  4.5 1.5]
 [6.3 3.4 5.6 2.4]
 [6.3 3.3 6.  2.5]
 [4.5 2.3 1.

<h3>Specify the training models (i.e. SVC, KNN, GNB)</h3>

In [15]:
# SVC_model = svm.SVC()
SVC_model = SVC(gamma='scale')
# KNN model requires you to specify n_neighbors,
# the number of points the classifier will look at to determine what class a new point belongs to
KNN_model = KNeighborsClassifier(n_neighbors=5)
# KNN_model = KNeighborsClassifier(algorithm='scale', leaf_size=30, metric='minkowski',
#                     metric_params=None, n_jobs=None, n_neighbors=5, p=2)
# Gaussian NB model
GNB_model = GaussianNB()

<h3>Fittig the models</h3>

In [16]:
SVC_model.fit(X_train, y_train)
KNN_model.fit(X_train, y_train)
GNB_model.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

<h3>Training completed. Use models to predict and store the outcome</h3>

In [17]:
SVC_prediction = SVC_model.predict(X_test)
KNN_prediction = KNN_model.predict(X_test)
GNB_prediction = GNB_model.predict(X_test)

<h3>Model Evaluation Using Accuracy Score and Confusion Matrix</h3>

In [18]:
# Accuracy score is the simplest way to evaluate
print(accuracy_score(SVC_prediction, y_test))
print(accuracy_score(KNN_prediction, y_test))
print(accuracy_score(GNB_prediction, y_test))
# But Confusion Matrix and Classification Report give more details about performance
print(confusion_matrix(SVC_prediction, y_test))
print(classification_report(KNN_prediction, y_test))
print(accuracy_score(GNB_prediction, y_test))

0.9333333333333333
0.9666666666666667
0.9
[[ 7  0  0]
 [ 0 10  1]
 [ 0  1 11]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       1.00      0.92      0.96        12
 Iris-virginica       0.92      1.00      0.96        11

       accuracy                           0.97        30
      macro avg       0.97      0.97      0.97        30
   weighted avg       0.97      0.97      0.97        30

0.9


<h3>Accuracy score of each model</h3>

SVC accuracy: 0.9333333333333333
KNN accuracy: 0.9666666666666667
GNB accuracy: 0.9